Websites/perf.webkit.org/public/v3/models/test-parameter.js

```diff
@@ -0,0 +1,212 @@
+            this.#testParameterMap.set(item.parameter, entry);
+            (item.parameter.type === 'build' ? this.#buildTypeParameters : this.#testTypeParameters).push(item.parameter);
+        }
+    }
+    get buildTypeParameters() { return this.#buildTypeParameters; }
+    get testTypeParameters() { return this.#testTypeParameters; }
+    get parameters() { return [...this.#buildTypeParameters, ...this.#testTypeParameters]; }
+    asPayload() { return this._createPayloadForParameters(this.parameters);}
+    hasSameBuildParametersAs(other) { return this._checkEquals(other, 'buildTypeParameters'); }
+    equals(other) { return this._checkEquals(other, 'parameters'); }
+    get(parameter) { return this.#testParameterMap.get(parameter); }
+    valueForParameter(parameter) { return this.#testParameterMap.get(parameter)?.value; }
+    fileForParameter(parameter) { return this.#testParameterMap.get(parameter)?.file; }
+    valueForParameterName(parameterName) { return this.valueForParameter(TestParameter.findByName(parameterName)); }
+    fileForParameterName(parameterName) { return this.fileForParameter(TestParameter.findByName(parameterName)); }
+
+    _createPayloadForParameters(parameters)
+    {
+        const payload = {};
+        for (const parameter of parameters) {
+            if (!this.#testParameterMap.has(parameter))
+                continue;
+            const entry = {};
+            if (parameter.hasValue)
+                entry.value = this.valueForParameter(parameter);
+            if (parameter.hasFile)
+                entry.value = this.fileForParameter(parameter).id()
+            payload[parameter.id()] = entry;
+        }
+        return payload;
+    }
+
+    _checkEquals(other, parameterName)
+    {
+        const thisParameters = this[parameterName];
+        if (!other)
+            return !thisParameters.length;
+        console.assert(other instanceof TestParameterSet);
+        const otherParameters = other[parameterName];
+        if (thisParameters.length !== otherParameters.length)
+            return false;
+        return thisParameters.every(parameter => TestParameterSet.isValueEqual(this.get(parameter), other.get(parameter)));
+    }
+
+    static isValueEqual(a, b) {
+        if (a === b)
+            return true;
+
+        if (typeof a !== 'object' || typeof b !== 'object' || a === null || b === null)
+            return false;
+
+        const aKeys = Object.keys(a);
+        const bKeySet = new Set(Object.keys(b));
+
+        if (aKeys.length !== bKeySet.size)
+            return false;
+
+        return aKeys.every(key => bKeySet.has(key) && TestParameterSet.isValueEqual(a[key], b[key]));
+    }
+
+    static equals(a, b)
+    {
+        return a == b || a?.equals(b) || b?.equals(a);
+    }
+
+    static hasSameBuildParameters(a, b)
+    {
+        return a?.hasSameBuildParametersAs(b) || b?.hasSameBuildParametersAs(a) || a == b;
+    }
+}
+
+class CustomTestParameterSet {
```

Comment:
Little confused as to what the difference between `TestParameterSet` and `CustomTestParameterSet` is - is it that the former is intended to be used as predefined parameters in the backend, and the latter for the frontend? If this is the case, should `CustomTestParameterSet` inherit `TestParameterSet`?

Reply:
Yes, `CustomTestParameterSet` is for front end editing to craft payload sent to backend API.
`CustomTestParameterSet` cannot inherit `TestParameterSet` because `TestParameterSet` inherits `DataModelObject` which presents database entries (has `id` field).